Visibility Modifier Check, ignore annotated fields

alexkravin opened the pull request:
Based on Pull #584

Requested points to fix:
```
Some items to fix before merge:

1)

mIgnoredAnnotationFormat
please make it static final private.

2)

mIgnoreAnnotated
mIgnoredAnnotationPattern
too much class members, please keep only mIgnoredAnnotationPattern with default value from mIgnoredAnnotationFormat, get exp should be applied each time.

3)

throw new ConversionException("unable to parse " + aPattern, e);
rethrow to top, Check do not know what to do on failure, it is business of top level

4)

final DetailAST firstIgnoredAnnotation =
AnnotationUtility.containsMatchingAnnotation(aAST, getIgnoredAnnotationPatternRegex());
final boolean hasIgnoredAnnotation = mIgnoreAnnotated
&& firstIgnoredAnnotation != null;
please make it like a function and move after " !"private" " condition , the easiest conditions go first.

5)

public static DetailAST containsMatchingAnnotation(final DetailAST aAST,
Pattern aAnnotationPattern)
please move that method to check as static function. Changes to API package is not allowed. I am sorry that we have utils in API, we will fix that soon.
Additionally please use 1 return point from method.

6)

throw new NullPointerException("the aAST is null");
please never throw NPE, use IllegalAgrument
```

1) Done

2) Only 2 fields are remained

3) Rethrowing to top

4) This code chunk is in separate boolean method and does not complex condition

5, 6) private static in Check's class, has one return point, no throwing of NPE, IllegalArgument used instead

P.S.

Something strange with coverage report as introduced code is fully-covered except for throwing exceptions

romani commented:
1)  Please update short description of javadoc (first paragraph)

2) 
> ignoredAnnotationPattern

make it "ignoreAnnotationPattern", we use the same name notation in other places.
please udpate whole code and all variables and methods.

3)
> containsMatchingAnnotation(variableDef, getIgnoredAnnotationPatternRegex());

please use field for second argument.

4) 
> 
if (ast == null) {
throw new IllegalArgumentException("The ast is null");
}

 if (annotationPattern == null) {
 throw new IllegalArgumentException("The annotationPattern is null");
 }

please remove that checks, till it is private method you do not do any good by such validations , NullPointerException and IllegalArgumentException are both mean development error, and processed by your check in the same way. No benefits at all by custom validation and exception usage here.

alexkravin commented:
1) Updated javadoc/xdoc

2) Updated

3) Ok

4) Removed these Checks, I thought they're useless even at first time, but you comment in previous PR 'chage NPE to IllegalArgument' has misled me

romani commented:
please make ignoreAnnotationPattern as list of Annotation canonical names, as that check already process imports, lets do type detection more exact.

alexkravin commented:
Does it have any sense?

I agree that class names can lead to ambiguity in some cases, but it's not too frequent, but ambiguity between annotations  is too weird to take care about it

romani commented:
I though a lot about that, yes it make sence.
Users are creazy with a java, and as it is compilable it is a correct code. We shoyld not produce false positives And we already play with imports in that check. It right now looks strange why one option is strict and another is not.

alexkravin commented:
Ok, I added support of ignore annotations list by the same way as immutable types are considered (canonical names, short names and dependence from imports)

romani commented:
1)
> property name=&quot;ignoreAnnotationPattern&quot;

No ignoreAnnotationPattern should be in changes.
Please recheck that xdoc have all changes from javadoc.

2)
>  \+ \* \<ul\>
\+ \* \<li\>org.junit.rule\</li\>

Rule

3)
> public void beginTree(DetailAST rootAst)

please do all processing for Classes and then for Annotations, do not mix logic.

alexkravin commented:
1) Updated javadoc/xdoc

2) Ok

3) Separated the logic of class and annotation names processing

romani commented:
1) I do not see TemporaryFolder example in xdoc

2) in code exsmples please make a case for short name of annotation.

alexkravin commented:
1, 2) Extended examples in xdoc, also added examples of fields annotated with short annotations name.

romani commented:
1) 
> @com.annotation.CustomAnnotation
\* String customAnnotated; // No warning

I wanted from you code like 
```
@com.annotation.CustomAnnotation
String customAnnotated; // No warning
@CustomAnnotation
String customAnnotated; // No warning
```

To show user that you detect all forms of type referencing.

2) 
> passing fields annotated with short annotation name

please extend existing example with code that shows you will detect all annotation with same name but different packages.

alexkravin commented:
1) Updated code example

2) Extended this code example by showing no violation on field annotated by annotation's canonical name with different package but matching short name, added comment describing this situation near the example in javadoc/xdoc

romani commented:
Please review my comments over the code.

alexkravin commented:
Updated according to comments.

romani commented:
merged as FF
